packages/runtime/container-runtime/src/blobAggregationStorage.ts

```diff
@@ -0,0 +1,365 @@
+        // that starts with this.aggregatedBlobName
+        for (const key of Object.keys(snapshot.blobs)) {
+            if (!key.startsWith(this.aggregatedBlobName)) { continue; }
+            const blobId = snapshot.blobs[key];
+            if (blobId !== undefined) {
+                const blob = await this.getBlob(blobId, snapshot);
+                for (const [path, value] of BlobAggregator.load(blob)) {
+                    const id = this.getNextVirtualId();
+                    this.setBlob(id, snapshot, value);
+                    const pathSplit = path.split("/");
+                    let subTree = snapshot;
+                    for (const subPath of pathSplit.slice(0, pathSplit.length - 1)) {
+                        if (subTree.trees[subPath] === undefined) {
+                            subTree.trees[subPath] = { blobs: {}, commits: {}, trees: {}};
+                        }
+                        subTree = subTree.trees[subPath];
+                    }
+                    const blobName = pathSplit[pathSplit.length - 1];
+                    assert(subTree.blobs[blobName] === undefined);
+                    subTree.blobs[blobName] = id;
+                }
+                // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+                delete snapshot.blobs[this.aggregatedBlobName];
+            }
+        }
+    }
+}
+
+/*
+ * Snapshot extractor class that works in place, i.e. patches snapshot that has
+ * blob content in ISnapshotTree.blobs itself, not in storage.
+ * As result, it implements reading and writing of blobs to/from snapshot itself.
+ * It follows existing pattern that mixes concerns - ISnapshotTree.blobs is used for two
+ * purposes:
+ * 1. map path name to blob ID
+ * 2. map blob ID to blob content
+ * #2 is what storage (IDocumentStorageService) is for, but in places where we do not have it
+ * (like loading serialized earlier draft content), blob content is put directly into snapshot.
+ * Ideally this should be fixed by using BlobCacheStorageService or something similar and
+ * fixing existing flows to allow switching of storage.
+ */
+class SnapshotExtractorInPlace extends SnapshotExtractor {
+    public async getBlob(id: string, tree: ISnapshotTree): Promise<ArrayBufferLike> {
+        const blob = tree.blobs[id];
+        assert(blob !== undefined);
+        return stringToBuffer(blob, "base64");
+    }
+
+    public setBlob(id: string, tree: ISnapshotTree, content: string) {
+        assert(tree.blobs[id] === undefined);
+        tree.blobs[id] = fromUtf8ToBase64(content);
+    }
+}
+
+/*
+ * Snapshot packer and extractor.
+ * When summary is written it will find and aggregate small blobs into bigger blobs
+ * When snapshot is read, it will unpack aggregated blobs and provide them transparently to caller.
+ */
+export class BlobAggregatorStorage extends SnapshotExtractor implements IDocumentStorageService {
```

Comment:
fluid-fetch integration will be necessary.

Reply:
It's good point, though I wonder - where does it stop? I.e. this "interpretation" layer that is on top of storage.
fluid-fetch tool already cracks some things, so probably it's fine, but it feels like it should be separate tool to give storage representation some "meaning".
Not pushing back, just saying that it adds some complexity and mixing concerns.

Reply:
That's very true, although fluid-fetch probably should able to report on various layer interpretation.

Reply:
Opened https://github.com/microsoft/FluidFramework/issues/5167 to do fast follow up - it needs to be done before I enable write part of this change (which should happen in ~2 releases)